proxy/envoy/config.go

```diff
@@ -303,9 +303,11 @@ func buildInboundFilters(instances []*model.ServiceInstance) (HTTPRouteConfigs,
 			http.VirtualHosts = append(http.VirtualHosts, host)
 		case model.ProtocolTCP:
 			cluster := buildInboundCluster(service.Hostname, endpoint.Port, port.Protocol)
-			route := buildTCPRoute(cluster, endpoint.Address, endpoint.Port)
 			config := tcpConfigs.EnsurePort(port.Port)
-			config.Routes = append(config.Routes, route)
+			config.Routes = append(config.Routes,
+				buildTCPRoute(cluster, []string{endpoint.Address}, endpoint.Port),
+				buildTCPRoute(cluster, []string{service.Address, endpoint.Address}, port.Port),
```

Comment:
It might be that you have both 90 and 9090 and they get mixed up (one port receives another port's traffic). I'd expect `inbound:9090` cluster to be only on listener 9090. There's no need to use destination port setting since it's known to be 9090. So a TCP service with port 90, and target port 9090, should have a TCP listener on port 9090 that sends to cluster `inbound:9090` and a listener on port 90 that sends to cluster `outbound:.....:90`